### DISCUSSION TITLE
Behavior change in assignment of 'tag' property of Token (spaCy 2.3.5 vs 3.0.5)

### DISCUSSION THREAD

**Discussion by nrodnova** [explosion/spaCy#7880] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour

In spacy 2, an assignment of the granular tag `token.tag_` would also change `token.pos` to a corresponding part of speach. In spacy 3 this behavior disappeared. And while a work-around is not hard, I think it was the right behavior. I hope you can put it back.

### Code fragment
```
nlp = spacy.load('en_core_web_sm')
txt = "We are running short of cooking oil."
doc = nlp(txt)
token = doc[5]
print("Token: %s, Tag: %s, POS: %s" % (token, token.tag_, token.pos_))
token.tag_ = 'JJ'
print("Token: %s, Tag: %s, POS: %s" % (token, token.tag_, token.pos_))
```
### Spacy 2 output:
```
Token: cooking, Tag: NN, POS: NOUN
Token: cooking, Tag: JJ, POS: ADJ
```
### Spacy 3 output:
```
Token: cooking, Tag: VBG, POS: VERB
Token: cooking, Tag: JJ, POS: VERB
```
As you can see, the second lines of both versions differ, in spacy 3 we how have a VERB with tag JJ.

The difference is coming from `Token`'s `tag` property. 
In spacy 2:
```
    property tag:
        """RETURNS (uint64): ID of fine-grained part-of-speech tag."""
        def __get__(self):
            return self.c.tag

        def __set__(self, attr_t tag):
            self.vocab.morphology.assign_tag(self.c, tag)

```
In spacy 3:
```
property tag:
        """RETURNS (uint64): ID of fine-grained part-of-speech tag."""
        def __get__(self):
            return self.c.tag

        def __set__(self, attr_t tag):
            self.c.tag = tag
```
## Info about spaCy

- **spaCy version:** 3.0.5
- **Platform:** Darwin-18.7.0-x86_64-i386-64bit
- **Python version:** 3.7.3
- **Pipelines:** en_core_web_lg (3.0.0), en_core_web_md (3.0.0), en_core_web_sm (3.0.0), en_core_web_trf (3.0.0)

**Comment by nrodnova** [comment]
Actually, I am seeing now that tag_map.py disappeared in spacy 3. Is this information stored somewhere else?

**Comment by polm** [comment, marked as answer]
Remapping the tags to UPOS happens in the [AttributeRuler](https://spacy.io/api/attributeruler) now. The remapping doesn't happen automatically on assignment, so you need to run that part of the pipeline again if you want to update the UPOS tags. Here's an example that does that:

```
import spacy

nlp = spacy.load("en_core_web_sm")
text = "I am cheese."
doc = nlp(text)
doc[2].tag_ = "VBG" # pretend "cheese" is a gerund

# no update here, pos is still NOUN
for tok in doc:
    print(tok, tok.tag_, tok.pos_, sep="\t")
print("-----")

aruler = nlp.get_pipe("attribute_ruler")
doc = aruler(doc)

# now it's updated and it's a VERB
for tok in doc:
    print(tok, tok.tag_, tok.pos_, sep="\t")
```

Note that the AttributeRuler can still [load tag maps](https://spacy.io/api/attributeruler#load_from_tag_map) if you have a custom one you'd like to use. 

Triggering a pipeline component on attribute assignment would be complicated and isn't something we're likely to do by default. That said, could you tell us more about your use case? Updating POS tags manually after the tagger but then using the UPOS tags they imply isn't something I've seen before.

**Reply by nrodnova** [reply]
Thank you @polm, this is helpful. I agree that triggering a pipeline component on assignment doesn't make sense. My use case is pretty simple. I work with medical texts and I identified about 2K frequent words that the tagger gets wrong. So I just overwrite those (on the granular level), because I don't have resources at the moment to annotate and re-train the tagger. Later, in some matcher rules, I occasionally use POS vs granular tags, when I simply need to match, say, an adjective.

**Reply by polm** [reply]
Ah, thanks for explaining your use case, that makes a lot of sense. In that case you can feed rules to the AttributeRuler to correct those tags and have everything updated as part of your pipeline - take a look at [the docs](https://spacy.io/usage/linguistic-features/#mappings-exceptions), where there's an example that fixes part of speech tags for the band "The Who".

**Reply by adrianeboyd** [reply]
As a note: if you want to have the current tag map rules in the attribute ruler set the final POS for you, you'll need to put your tag corrections in a separate attribute ruler component that's inserted between the tagger and existing attribute ruler. To be fast, the attribute ruler only runs one matching pass over the doc, so cascading patterns where one rule's input depends on another rule's output don't work. The internal matcher within the attribute ruler only matches on the original doc annotations, not on any intermediate/partial modifications.

The alternative with one attribute ruler is to have your custom tag correction patterns set both TAG and POS. This is a bit redundant, but if it's straightforward for you to write/maintain the rules with both TAG and POS, it will probably also be slightly faster because the matcher only runs once.

**Reply by nrodnova** [reply]
Thank you, @adrianeboyd, that is what I ended up doing - meaning, assigning both TAG and POS. I couldn't find the tag mapping anywhere in spacy 3 though (what used to be tag_map.py) I copied one from v2 into my local project, but I'd rather import and use whatever spaCy uses, just to keep it clean.

**Reply by adrianeboyd** [reply]
The tag map rules are included in the provided `attribute_ruler`, but there are other rules included, too, so there's no clean separation between what used to be tag map rules and what came from morph rules or elsewhere. I guess you can sift out all the rules that set `POS`. It won't be exactly equivalent to the v2 tag maps because we've added some rules that use `DEP` to improve the mapping for some forms of `to be`.
